File: src/playhaus-api/internal/pubquizr/round_one.go
package pubquizr

// Round 1: who is being asked, and who reads next.
//
// The round is a hot seat. The quizmaster reads a question to whoever is in it; if
// they miss it, it passes on round the table -- still the same question, still the
// same reader -- until somebody takes it or everybody but the reader has had their
// one go. Taking it puts you in the seat for the next question too, which is how a
// player holds it: the reading only moves when a question dies unanswered.
//
// The seat is therefore not something that can be worked out from who is reading,
// and `Session.HotSeat` is where it lives. Everything here is arithmetic around it.
//
// Kept apart from the service so the ordering can be read, and tested, without a
// database in the way.

// AnsweringSeat is whose turn it is to answer, given where the question started and
// how many seats have already tried it.
//
// hotSeat is the seat the question was first asked to. attempts is the number of
// goes already had: 0 is that seat, 1 is the next one along, and so on. Returns -1
// once every seat but the reader has had a go, which is the question running out
// rather than a seat to ask.
//
// The walk skips the quizmaster rather than stopping at them. A question can now
// start anywhere, so "back at the reader" is a seat to step over on the way round
// and no longer says anything about how much of the table is left.
func AnsweringSeat(quizMasterSeat, hotSeat, attempts, players int) int {
	if players <= 1 || attempts < 0 {
		return -1
	}
	// Everybody except the reader gets at most one go, so there are players-1 of
	// them and the next attempt after that is nobody.
	if attempts >= players-1 {
		return -1
	}

	// The seats that can be asked are every seat but the reader, in table order from
	// their left. Working along that ring rather than in seat numbers is what makes
	// skipping the quizmaster fall out instead of needing a loop that steps over
	// them.
	first := wrap(quizMasterSeat+1, players)

	// Where the hot seat sits in that ring. A hot seat that is somehow the reader --
	// or one never set -- lands on players-1, which is 0 once it is taken round the
	// ring, so the question starts to the reader's left the way it always used to.
	start := wrap(hotSeat-first, players)

	return wrap(first+(start+attempts)%(players-1), players)
}

// wrap is n modulo size, always non-negative. Go's % keeps the sign of its left
// operand, which is not what walking backwards round a table wants.
func wrap(n, size int) int {
	return ((n % size) + size) % size
}

// QuestionsIn is how many questions a round of this session holds.
//
// Read off the deal rather than off the quiz: what this table plays was frozen when
// the game started, and the quiz behind it may hold more than they were dealt.
func (s *Session) QuestionsInRound(round int) int {
	count := 0

	for _, question := range s.Questions {
		if question.Round == round {
			count++
		}
	}

	return count
}

// QuestionAt is the dealt question in one slot, or nil when the round is over.
func (s *Session) QuestionAt(round, position int) *SessionQuestion {
	for i := range s.Questions {
		if s.Questions[i].Round == round && s.Questions[i].Position == position {
			return &s.Questions[i]
		}
	}

	return nil
}

// PlayerAt is whoever is sitting in one seat, or nil for a seat that is not at this
// table.
func (s *Session) PlayerAt(seat int) *SessionPlayer {
	for i := range s.Players {
		if s.Players[i].Seat == seat {
			return &s.Players[i]
		}
	}

	return nil
}

// HotSeatOrFirst is where the current question started.
//
// The stored seat, except for the two cases where it cannot be believed: a session
// dealt before the column existed carries -1, and a seat that has ended up being the
// reader's own is a table that could not be asked anything. Both fall back to the
// seat on the quizmaster's left, which is where a fresh question starts anyway.
func (s *Session) HotSeatOrFirst() int {
	players := len(s.Players)
	if players == 0 {
		return -1
	}

	if s.HotSeat < 0 || s.HotSeat >= players || s.HotSeat == s.QuizMasterSeat {
		return wrap(s.QuizMasterSeat+1, players)
	}

	return s.HotSeat
}

// CurrentAnsweringSeat is whose turn it is to answer right now, or -1 when nobody
// is being asked anything -- a finished session, or a round this file has no say
// over.
//
// attempts is what the store counted for the current question.
func (s *Session) CurrentAnsweringSeat(attempts int) int {
	if s.Status != SessionInProgress || s.CurrentRound != RoundOpen {
		return -1
	}
	if s.QuestionAt(s.CurrentRound, s.CurrentPosition) == nil {
		return -1
	}

	return AnsweringSeat(s.QuizMasterSeat, s.HotSeatOrFirst(), attempts, len(s.Players))
}

// ReaderFor is who reads to one seat: the player on their right, which is the seat
// before them in table order.
//
// The round's whole ordering, in one line. A question is read by the neighbour of
// whoever it opens on, so naming the seat that starts names the reader too -- and
// every way the game moves on works by moving the start and letting the reading
// follow it.
func ReaderFor(seat, players int) int {
	if players <= 0 {
		return -1
	}

	return wrap(seat-1, players)
}

// OpenOn puts the next question on one seat and the reading on the seat to its
// right.
//
// The only thing that should ever write either of the two columns, because they are
// one fact: a hot seat with somebody other than its right-hand neighbour reading to
// it is a table nobody at it could describe.
func (s *Session) OpenOn(seat int) {
	players := len(s.Players)
	if players == 0 {
		return
	}

	s.HotSeat = wrap(seat, players)
	s.QuizMasterSeat = ReaderFor(s.HotSeat, players)
}

// LowestScoringSeat is whoever has the fewest points, ties going to whoever sits
// nearest the head of the table.
//
// Every round but the first and the finale opens on them: the round starts with the
// person who most needs it to. Ties break on the seat rather than at random so the
// answer is the same one twice, which is what lets a table argue with it.
func (s *Session) LowestScoringSeat() int {
	seat, score := -1, 0

	for _, player := range s.Players {
		if seat < 0 || player.Score < score || (player.Score == score && player.Seat < seat) {
			seat, score = player.Seat, player.Score
		}
	}

	return seat
}

File: src/playhaus-app/src/features/pubquizr/round-one.ts
import { Brand } from "@/constants/theme";
import { avatarColorById, type AvatarColor } from "@/features/settings/profile";
import type { QuizDetail, QuizQuestion } from "./pubquizr-quizzes";
import type { QuizSession, QuizSessionPlayer, QuizSessionQuestion } from "./pubquizr-sessions";

/**
 * Round 1, as the screen needs it.
 *
 * The session says *which* questions are being played and how it is going; the quiz
 * says what they actually are. Neither is much use alone — the session carries a
 * `questionId` and no words, and the quiz has no idea whose turn it is — so everything
 * here is the join between the two, worked out once and handed to the components as
 * things they can draw without knowing where any of it came from.
 *
 * The round itself is a hot seat: taking a question keeps you in it for the next one,
 * the reading follows the seat round the table — you are always read to by the player
 * on your right — and only every second question is worth a point. None of that is
 * decided here; it all comes back from the server. But it has to be *said* on screen
 * before a button is pressed, which is what `scoresAt`, `nextUpAfter` and the run on
 * `RoundOneTurn` below are for.
 */

/** The round the whole of this file is about. Kept in step with `RoundOpen` in Go. */
export const ROUND_OPEN = 1;

/**
 * How often a question in this round is worth a point: every second one.
 *
 * Kept in step with `OpenScoresEvery` in `rules.go`, and duplicated here for the same
 * reason `nextUpAfter` below duplicates the pass-along arithmetic — this is only ever
 * used to *say* what the question on screen is worth. The server still decides, and
 * the score comes back from it.
 */
export const OPEN_SCORES_EVERY = 2;

/** Whether this question pays out, given its 1-based number in the round. */
export function scoresAt(questionNumber: number): boolean {
    return questionNumber % OPEN_SCORES_EVERY === 0;
}

export interface Seat {
    seat: number
    name: string
    score: number
    /** Two letters for the swatch — `SA` for Sanne. */
    initials: string
    swatch: AvatarColor
}

export interface RoundOneTurn {
    /** The dealt question being played, which is what a verdict has to name. */
    dealt: QuizSessionQuestion
    /** What it actually says, out of the quiz. */
    question: QuizQuestion
    /** The answer to read off the back of the card. */
    answer: string
    /** Wordings that also count. Never the headline answer. */
    aliases: string[]
    /** Whoever is reading it out. */
    quizmaster: Seat
    /** Whoever is being asked right now. */
    answering: Seat
    /**
     * How many questions in a row `answering` has taken, and 0 when they have taken
     * none — a question that has passed on to them is a fresh start.
     *
     * The number behind the rule the round is built on: get it right and you are asked
     * the next one too. Straight off the server, which counts it.
     */
    run: number
    /** Who gets it if this one is wrong, or null when the question is on its last seat. */
    nextUp: Seat | null
    /** 1-based, for "question 3 of 20". */
    number: number
    total: number
    /** Whether taking this one is worth a point, or only worth staying in for. */
    scoring: boolean
}

/**
 * The two letters on a seat's swatch.
 *
 * Spread rather than sliced, because a name starting with an accented pair cut by
 * index comes out as half a glyph — the same reason `initialsFor` in `quiz-shelf.ts`
 * does it, which this deliberately mirrors rather than imports: that one is about a
 * two-word quiz title, and a person is one word whose first two letters are wanted.
 */
export function initialsOf(name: string): string {
    const letters = [...name.trim()].filter(character => character.trim() !== '');

    return letters.slice(0, 2).join('').toUpperCase() || '?';
}

export function seatOf(player: QuizSessionPlayer): Seat {
    return {
        seat: player.seat,
        name: player.name,
        score: player.score,
        initials: initialsOf(player.name),
        swatch: avatarColorById(player.color)
    };
}

/** Everybody at the table, in seating order. */
export function seatsOf(session: QuizSession): Seat[] {
    return [...session.players].sort((a, b) => a.seat - b.seat).map(seatOf);
}

/** The standings, best first, with ties left in seating order. */
export function standingsOf(session: QuizSession): Seat[] {
    return seatsOf(session).sort((a, b) => b.score - a.score || a.seat - b.seat);
}

/**
 * What is on screen right now, or null when round 1 is not what is being played.
 *
 * Null covers every "there is no turn" case at once — the round has moved on, the
 * session is finished, nobody is being asked — so the screen has one question to ask
 * rather than four.
 */
export function turnOf(session: QuizSession, quiz: QuizDetail): RoundOneTurn | null {
    if (session.status !== 'in_progress') return null;
    if (session.currentRound !== ROUND_OPEN) return null;
    if (session.answeringSeat === null) return null;

    const dealt = session.questions.find(
        question => question.round === session.currentRound
            && question.position === session.currentPosition
    );
    if (dealt === undefined) return null;

    const question = quiz.rounds
        .flatMap(round => round.questions)
        .find(candidate => candidate.id === dealt.questionId);
    if (question === undefined) return null;

    const seats = seatsOf(session);
    const quizmaster = seats.find(seat => seat.seat === session.quizMasterSeat);
    const answering = seats.find(seat => seat.seat === session.answeringSeat);
    if (quizmaster === undefined || answering === undefined) return null;

    // The headline answer and the wordings that also count. An open question carries
    // one real answer and any number of aliases behind it.
    const answers = question.answers.filter(answer => answer.alias !== true);
    const aliases = question.answers.filter(answer => answer.alias === true);

    return {
        dealt,
        question,
        answer: answers.map(answer => answer.text).join(' / '),
        aliases: aliases.map(answer => answer.text),
        quizmaster,
        answering,
        // A run belongs to whoever is *holding* the seat. Once a question has passed
        // along, the person now being asked has taken nothing yet, so there is no run
        // of theirs to put on the board.
        run: session.answeringSeat === session.hotSeat ? session.hotSeatRun : 0,
        nextUp: nextUpAfter(session, seats),
        number: session.currentPosition + 1,
        total: session.questions.filter(candidate => candidate.round === ROUND_OPEN).length,
        scoring: scoresAt(session.currentPosition + 1)
    };
}

/**
 * Who a wrong answer would pass the question to, or null when there is nobody left.
 *
 * The same arithmetic the server does in `round_one.go`, and it is duplicated on
 * purpose: this is only ever used to *say* what the wrong button will do before it is
 * pressed. The server is still the one that decides, and the next screen comes back
 * from it — so a disagreement here is a misleading line of small print rather than a
 * point going to the wrong person.
 *
 * The reader is stepped over rather than stopped at. A question no longer has to start
 * on their left, so landing on them says nothing about how much of the table is left;
 * what ends the question is arriving back at the seat it opened on, which is the point
 * at which everybody but the reader has had their one go.
 */
function nextUpAfter(session: QuizSession, seats: Seat[]): Seat | null {
    if (session.answeringSeat === null) return null;
    if (seats.length <= 1) return null;

    let next = (session.answeringSeat + 1) % seats.length;
    if (next === session.quizMasterSeat) {
        next = (next + 1) % seats.length;
    }

    // All the way round to where it started: everybody else has already said no.
    if (next === session.hotSeat) return null;

    return seats.find(seat => seat.seat === next) ?? null;
}

/** How a hand-off screen is painted: one fill, and the ink that stays readable on it. */
export interface HandoffTone {
    fill: string
    /** Headline text. */
    ink: string
    /** The same ink stepped back, for the step label and the explanation line. */
    muted: string
}

/**
 * The five fills a hand-off can wear, in the order they come round.
 *
 * Lemon alone was a screen you stop seeing: the hand-off looks the same every time, so
 * after a few questions it stops registering as a new one. A different colour each turn
 * is the cheapest way to make "this is a new screen, stop and read it" land, which is
 * the only job that screen has.
 *
 * Black on every fill but the blue, which is the one that cannot carry it: ink on
 * #3B4DF0 is 3.2:1, so the 38px name would just about pass and the 15px line under it
 * would not come close, and a hand-off whose instructions cannot be read is the one
 * thing this screen cannot afford. Orange looks like it wants paper and does not —
 * ink on it is 6.2:1 against paper's 3.0:1.
 *
 * The muted alphas are the weakest each fill can carry and still clear 4.5:1 for body
 * text, rounded up: 0.76 on orange and 0.88 on the blue are doing real work, and the
 * three light fills are held at 0.7 for the sake of looking like one family.
 */
const HANDOFF_TONES: HandoffTone[] = [
    { fill: Brand.primary,   ink: Brand.ink,          muted: 'rgba(15, 13, 18, 0.78)' },
    { fill: Brand.lemon,     ink: Brand.ink,          muted: 'rgba(15, 13, 18, 0.7)' },
    { fill: Brand.mint,      ink: Brand.ink,          muted: 'rgba(15, 13, 18, 0.7)' },
    { fill: Brand.blush,     ink: Brand.ink,          muted: 'rgba(15, 13, 18, 0.7)' },
    { fill: Brand.secondary, ink: Brand.textOnAccent, muted: 'rgba(254, 251, 248, 0.88)' }
];

/**
 * Which fill this turn's hand-off wears.
 *
 * Keyed off the question number rather than off who is taking the phone, so the colour
 * means "the screen changed" rather than "this person". Two hand-offs in a row are
 * always different, which is the whole point.
 */
export function handoffToneFor(questionNumber: number): HandoffTone {
    // 1-based, so the first question of the round gets the first tone.
    const index = (questionNumber - 1) % HANDOFF_TONES.length;

    return HANDOFF_TONES[(index + HANDOFF_TONES.length) % HANDOFF_TONES.length];
}
